File: sdks/python/tests/kinds.py
"""Small components of every other kind, for the SDK's own tests."""

from __future__ import annotations

from dataclasses import dataclass, replace
from datetime import timedelta

from ankka import DONE, Done, ErrorCode, ReadOnlyEffect, command, json_codec, query
from ankka.consumer import Consumer
from ankka.effects.consumer import ConsumerEffect
from ankka.effects.key_value import KeyValueEffect, KeyValueReadOnlyEffect
from ankka.effects.timed_action import TimedActionEffect
from ankka.effects.view import ViewEffect
from ankka.effects.workflow import WorkflowEffect, WorkflowReadOnlyEffect, WorkflowStepEffect
from ankka.key_value_entity import KeyValueEntity
from ankka.timed_action import TimedAction, action
from ankka.view import View
from ankka.workflow import Recovery, StepSettings, WorkflowSettings, Workflow, step
from tests.counter import CounterEntity, CounterEvent, Incremented, Noted


@dataclass(frozen=True)
class Profile:
    name: str
    visits: int


class ProfileEntity(KeyValueEntity[Profile]):
    component_id = "profile"
    state_codec = json_codec(Profile, "profile")

    def empty_state(self) -> Profile:
        return Profile("", 0)

    @command("set")
    def set_name(self, name: str) -> KeyValueEffect[Profile, Done]:
        if not name:
            return self.effects.error("a name is required", ErrorCode.BAD_REQUEST)
        return self.effects.update_state(replace(self.state, name=name)).then_reply(lambda _: DONE)

    @command("visit")
    def visit(self) -> KeyValueEffect[Profile, int]:
        return self.effects.update_state(replace(self.state, visits=self.state.visits + 1)).then_reply(lambda p: p.visits)

    @query("get")
    def get(self) -> KeyValueReadOnlyEffect[Profile, Profile]:
        return self.effects.reply(self.state)

    @command("delete")
    def delete(self) -> KeyValueEffect[Profile, Done]:
        return self.effects.delete_entity().then_reply(lambda _: DONE)


@dataclass(frozen=True)
class Checkout:
    cart_id: str
    status: str
    charged: int


@dataclass(frozen=True)
class Charge:
    amount: int
    fail: bool = False


class CheckoutWorkflow(Workflow[Checkout]):
    component_id = "checkout"
    state_codec = json_codec(Checkout, "checkout")
    settings = WorkflowSettings(
        default_step_timeout=timedelta(seconds=5),
        steps={"charge": StepSettings(timeout=timedelta(seconds=2), recovery=Recovery(max_retries=1, failover_to="compensate"))},
    )

    def empty_state(self) -> Checkout:
        return Checkout(self.entity_id, "new", 0)

    @command("start")
    def start(self, charge: Charge) -> WorkflowEffect[Checkout, Done]:
        if self.state.status != "new":
            return self.effects.error("already started", ErrorCode.CONFLICT)
        return self.effects.update_state(replace(self.state, status="reserving")).then_transition_to("reserve", charge).then_reply(lambda _: DONE)

    @query("status")
    def status(self) -> WorkflowReadOnlyEffect[Checkout, str]:
        return self.effects.reply(self.state.status)

    @step("reserve")
    def reserve(self, charge: Charge) -> WorkflowStepEffect[Checkout]:
        return self.step_effects.update_state(replace(self.state, status="reserved")).then_transition_to("charge", charge)

    @step("charge")
    async def charge(self, charge: Charge) -> WorkflowStepEffect[Checkout]:
        if charge.fail:
            return self.step_effects.update_state(replace(self.state, status="failed")).then_transition_to("compensate")
        return self.step_effects.update_state(replace(self.state, status="charged", charged=charge.amount)).then_end()

    @step("compensate")
    def compensate(self) -> WorkflowStepEffect[Checkout]:
        return self.step_effects.update_state(replace(self.state, status="compensated")).then_end()

    @step("wait")
    def wait(self) -> WorkflowStepEffect[Checkout]:
        return self.step_effects.pause(after=timedelta(seconds=1))


@dataclass(frozen=True)
class CounterRow:
    total: int
    notes: int


class CounterRows(View[CounterEvent, CounterRow]):
    component_id = "counter-rows"
    source = CounterEntity
    event_codec = CounterEntity.event_codec
    row_codec = json_codec(CounterRow, "counter-row")

    def on_change(self, event: CounterEvent) -> ViewEffect:
        current = self.row or CounterRow(0, 0)
        match event:
            case Incremented(by):
                if by == 999:
                    return self.effects.delete_row()
                return self.effects.update_row(replace(current, total=current.total + by))
            case Noted(_):
                return self.effects.update_row(replace(current, notes=current.notes + 1))
        return self.effects.ignore()


@dataclass(frozen=True)
class Alert:
    text: str


class BigIncrements(Consumer[CounterEvent, Alert]):
    component_id = "big-increments"
    source = CounterEntity
    message_codec = CounterEntity.event_codec
    produces_to = "alerts"
    out_codec = json_codec(Alert, "alert")

    def on_message(self, message: CounterEvent) -> ConsumerEffect:
        match message:
            case Incremented(by) if by >= 10:
                return self.effects.produce(Alert(f"+{by}"))
            case Incremented(_):
                return self.effects.done()
        return self.effects.ignore()


class Reminder(TimedAction):
    component_id = "reminder"

    @action("remind")
    def remind(self, who: str) -> TimedActionEffect:
        if who == "nobody":
            return self.effects.fail("nobody to remind", ErrorCode.NOT_FOUND)
        return self.effects.done()

File: sdks/typescript/test/fixtures/kinds.ts
// Fixtures for the other kinds: small components whose handlers cover each effect and need no sidecar.
import { KeyValueEntity } from "../../src/keyValueEntity.ts"
import { Workflow } from "../../src/workflow.ts"
import { Consumer } from "../../src/consumer.ts"
import { TimedAction } from "../../src/timedAction.ts"
import { Agent } from "../../src/agent.ts"
import { action, command, guardrail, query, step, stream, tool } from "../../src/handlers.ts"
import { jsonCodec } from "../../src/codec.ts"
import { Done, done, s, type Infer } from "../../src/schema.ts"
import { ErrorCode } from "../../src/effects/common.ts"
import { workflowSettings } from "../../src/effects/workflow.ts"
import { Duration } from "../../src/time.ts"

// ── key value ──

export const ProfileState = s.record("ProfileState", { name: s.string, visits: s.int })

export class Profile extends KeyValueEntity<Infer<typeof ProfileState>> {
  static readonly componentId = "profile"
  static readonly state = jsonCodec(ProfileState, "profile")
  static readonly handlers = {
    set: command("set", s.string, s.string, (p: Profile, name) => (name ? p.effects.updateState({ ...p.state, name, visits: p.state.visits + 1 }).thenReply((st) => `${st.name}#${st.visits}`) : p.effects.error("a name is needed"))),
    get: query("get", s.string, (p: Profile) => p.effects.reply(p.state.name || "none")),
    getState: query("get-state", ProfileState, (p: Profile) => p.effects.reply(p.state)),
    remove: command("delete", s.string, (p: Profile) => p.effects.deleteEntity().thenReply(() => "done")),
    expire: command("expire", Done, (p: Profile) => p.effects.updateState(p.state).expireAfter(Duration.ofSeconds(2)).thenReply(() => done)),
    boom: command("boom", Done, (): never => {
      throw new Error("kaboom")
    }),
  }
  emptyState() {
    return { name: "", visits: 0 }
  }
}

// ── workflow ──

export const Approval = s.record("Approval", { id: s.string, status: s.string, mode: s.string, note: s.option(s.string) })
export type Approval = Infer<typeof Approval>

export class ApprovalWorkflow extends Workflow<Approval> {
  static readonly componentId = "approval"
  static readonly state = jsonCodec(Approval, "approval")
  static readonly settings = workflowSettings({
    defaultStepTimeout: Duration.ofSeconds(5),
    steps: { check: { recovery: { maxRetries: 2, failoverTo: "finish" } } },
  })
  static readonly handlers = {
    start: command("start", s.string, Done, (w: ApprovalWorkflow, mode) =>
      w.state.status !== "new" ? w.effects.error(`already ${w.state.status}`, ErrorCode.Conflict) : w.effects.updateState({ ...w.state, status: "checking", mode }).thenTransitionTo("check").thenReply(() => done),
    ),
    status: query("status", s.string, (w: ApprovalWorkflow) => w.effects.reply(w.state.status)),
    approve: command("approve", Done, (w: ApprovalWorkflow) => w.effects.updateState({ ...w.state, status: "approving" }).thenTransitionTo("finish", "approved by hand").thenReply(() => done)),
  }
  static readonly steps = {
    check: step("check", (w: ApprovalWorkflow) => w.check()),
    slow: step("slow", async (w: ApprovalWorkflow) => {
      await new Promise((r) => setTimeout(r, 150))
      return w.stepEffects.updateState({ ...w.state, status: "slowed" }).thenEnd()
    }),
    finish: step("finish", s.string, (w: ApprovalWorkflow, note) => w.stepEffects.updateState({ ...w.state, status: "done", note }).thenEnd()),
  }
  emptyState(): Approval {
    return { id: this.entityId, status: "new", mode: "ok", note: null }
  }
  check() {
    switch (this.state.mode) {
      case "pause":
        return this.stepEffects.updateState({ ...this.state, status: "waiting" }).thenPause({ after: Duration.ofMillis(500), onTimeout: "finish", onTimeoutInput: "timed out" })
      case "fail":
        return this.stepEffects.fail("declined", ErrorCode.Forbidden)
      case "throw":
        throw new Error("check exploded")
      default:
        return this.stepEffects.updateState({ ...this.state, status: "checked" }).thenTransitionTo("finish", "all good")
    }
  }
}

// ── consumer, producing onward ──

export const Ping = s.record("Ping", { n: s.int })
export const Pong = s.record("Pong", { n: s.int, from: s.string })

export class Ponger extends Consumer<Infer<typeof Ping>, Infer<typeof Pong>> {
  static readonly componentId = "ponger"
  static readonly topic = "pings"
  static readonly message = jsonCodec(Ping, "ping")
  static readonly out = jsonCodec(Pong, "pong")
  static readonly producesTo = "pongs"
  onMessage(ping: Infer<typeof Ping>) {
    if (ping.n < 0) return this.effects.ignore()
    if (ping.n === 0) return this.effects.done()
    return this.effects.produce({ n: ping.n, from: this.subject }, { "x-ping": String(ping.n) })
  }
}

// ── timed action ──

export class Reminder extends TimedAction {
  static readonly componentId = "reminder"
  static readonly actions = {
    remind: action("remind", s.string, (r: Reminder, id) => (id === "bad" ? r.effects.fail("no such id", ErrorCode.NotFound) : r.effects.done())),
    ping: action("ping", (r: Reminder) => r.effects.done()),
  }
}

// ── agent ──

export const AddInput = s.record("AddInput", { a: s.int, b: s.int })
export const Sum = s.record("Sum", { sum: s.int })

export class Calculator extends Agent {
  static readonly componentId = "calculator"
  static readonly role = "adds numbers"
  static readonly maxToolCallSteps = 3
  static readonly tools = {
    add: tool("add", "Adds two whole numbers.", AddInput, (_c: Calculator, input) => {
      if (input.a > 1000) throw new Error("too big")
      return input.a + input.b
    }),
  }
  static readonly guardrails = {
    polite: guardrail("polite", (stage, text) => (text.includes("stupid") ? `${stage} was rude` : null)),
  }
  static readonly handlers = {
    ask: command("ask", s.string, s.string, (c: Calculator, q) => c.effects.systemMessage("You add numbers with the add tool.").userMessage(q).tools("add").guardrails("polite").thenReply()),
    askJson: command("ask-json", s.string, Sum, (c: Calculator, q) => c.effects.userMessage(q).tools("add").thenReplyJson<Infer<typeof Sum>>("{sum: number}")),
    forget: command("forget", s.string, s.string, (c: Calculator, q) => c.effects.userMessage(q).memory(false).thenReply()),
    refuse: command("refuse", s.string, s.string, (c: Calculator) => c.effects.error("not today", ErrorCode.Forbidden)),
    rogue: command("rogue", s.string, s.string, (c: Calculator, q) => c.effects.userMessage(q).tools("subtract").thenReply()),
    chat: stream("chat", s.string, (c: Calculator, q) => c.effects.userMessage(q).tools("add").thenReply()),
  }
}
